js/node_shared/unix.js

```diff
@@ -0,0 +1,348 @@
+   are not defined symbols in the produced JS.
+   To make sure we can still call those C functions (which happens in things
+   such as Bos, Time_limit, etc), we reimplement those fundamental Unix calls
+   in Javascript.
+   Note that this doesn't need to be a perfect reimplementation, but we can
+   get by with "good enough" behavior.
+ */
+
+const { spawnSync } = require("node:child_process");
+const os = require("os");
+const fs = require("fs");
+const process = require("process");
+
+globalThis.spawnSync = spawnSync;
+globalThis.fs = fs;
+globalThis.os = os;
+globalThis.process = process;
+
+// mapping pid to its own process
+globalThis.pidToProcessTable = new Map();
+// mapping pid of parent to its children
+globalThis.pidToChildrenTable = new Map();
+// There will be collisions if we ever allocate 20000 or more file descriptors
+// "naturally", that is, through `caml_sys_open`.
+// "Fake", because these file descriptors we produce are not real file descriptors,
+// but just indices corresponding to `unix_spawn` calls.
+// According to Austin: should be Ok, because node uses libuv
+// https://www.geeksforgeeks.org/libuv-in-node-js/
+globalThis.fakeFdThreshold = 20000;
+globalThis.fdCount = globalThis.fakeFdThreshold;
+globalThis.fdTable = new Map();
+
+/* Some magic happens here, in conjunction with unix_pipe.
+   Read its description for more.
+   The basic TL;DR is we generate fake file descriptors when piping or duping,
+   which we then associate to the output of this process, using Node's inherent
+   IO redirection capabilities.
+   Then, when reading from this "file descriptor", we just get the associated
+   output.
+ */
+//Provides: unix_spawn
+function unix_spawn(executable, args, optenv, usepath, redirect) {
+  // get rid of the mandatory tag and the first argument, which is the
+  // executable path
+  let argv = args.slice(2);
+
+  // These redirects are given to `unix_spawn` by Bos via `Unix.create_process_gen`,
+  // which gets them from Bos explicitly calling `pipe`.
+  // Via our convention in unix_pipe, these are not actually "real" file descriptors,
+  // but fresh identifiers we will use to map to the output of the corresponding
+  // child process.
+  // As such, let's just use the `pipe` setting for the stdio of this child process.
+  let output_fd = redirect[2];
+
+  /* In Javascript, there is no `execvp` and `execv` which
+     switch between searching in the PATH or not.
+     We will just run the vanilla `exec`, which should behave
+     as an `execvp`.
+     This might have slightly differing behavior than the
+     actual implementation.
+   */
+  if (optenv == 0) {
+    const child = globalThis.spawnSync(executable, argv, { stdio: "pipe" });
+
+    globalThis.pidToProcessTable.set(child.pid, child);
+    globalThis.fdTable.set(output_fd, child.stdout);
+
+    return child.pid;
+  }
+  let env = {};
+  for (const arg of optenv[1].slice(1)) {
+    const [key, value] = arg.split("=");
```

Comment:
`.split("=")` won't properly handle situations like `foo=bar=baz`, and `.split("=", 2)` will only give you `["foo", "bar"]`. We gotta use `.indexOf("=")` and slice out key and value ourselves